File: 224-basic-calculator/basic-calculator.cs
public class Solution {
    public int Calculate(string s) {
        int len = s.Length;
        int currentNumber = 0;
        int result = 0;
        int sign = 1;
        Stack<int> stack = new Stack<int>();
        
        for (int i = 0; i < len; i++) {
            char ch = s[i];
            
            if (char.IsDigit(ch)) {
                currentNumber = currentNumber * 10 + (ch - '0');
            } else if (ch == '+') {
                result += sign * currentNumber;
                currentNumber = 0;
                sign = 1;
            } else if (ch == '-') {
                result += sign * currentNumber;
                currentNumber = 0;
                sign = -1;
            } else if (ch == '(') {
                stack.Push(result);
                stack.Push(sign);
                result = 0;
                sign = 1;
            } else if (ch == ')') {
                result += sign * currentNumber;
                currentNumber = 0;
                result *= stack.Pop();
                result += stack.Pop();
            }
        }
        
        if (currentNumber != 0) {
            result += sign * currentNumber;
        }
        
        return result;
    }
}

File: 224-basic-calculator/basic-calculator.js
/**
*@param{string}s
*@return {number}
*/
var calculate=function(s){
    let stack=[], num=0, sign=1,result=0;
    for (let i=0; i<s.length; i++){
        let char=s[i];
        if(char >='0'&& char<='9'){num=num*10+parseInt(char);}
        else if(char ==='+'){result +=sign*num;sign=1;num=0;}
        else if(char==='-'){result +=sign*num; sign=-1;num=0;}
        else if(char==='('){stack.push(result);stack.push(sign);result=0; sign=1;}
        else if(char===')'){result +=sign*num; result *=stack.pop();result +=stack.pop();num=0;}
    }result +=sign *num; return result;
};
